validator/src/validation/email.rs

```diff
@@ -73,6 +42,68 @@ fn validate_domain_part(domain_part: &str) -> bool {
     }
 }
 
+pub trait ValidateEmail {
+    fn validate_email(&self) -> bool {
+        let val = self.to_email_string();
+
+        if val.is_empty() || !val.contains('@') {
+            return false;
+        }
+
+        let parts: Vec<&str> = val.rsplitn(2, '@').collect();
+        let user_part = parts[1];
+        let domain_part = parts[0];
+
+        // validate the length of each part of the email, BEFORE doing the regex
+        // according to RFC5321 the max length of the local part is 64 characters
+        // and the max length of the domain part is 255 characters
+        // https://datatracker.ietf.org/doc/html/rfc5321#section-4.5.3.1.1
+        if user_part.length() > 64 || domain_part.length() > 255 {
+            return false;
+        }
+
+        if !EMAIL_USER_RE.is_match(user_part) {
+            return false;
+        }
+
+        if !validate_domain_part(domain_part) {
+            // Still the possibility of an [IDN](https://en.wikipedia.org/wiki/Internationalized_domain_name)
+            return match domain_to_ascii(domain_part) {
+                Ok(d) => validate_domain_part(&d),
+                Err(_) => false,
+            };
+        }
+
+        true
+    }
+
+    fn to_email_string(&self) -> String;
```

Comment:
That implies allocation, it should return a `&str` instead

Reply:
Returning a `&str` is impossible if I want to return a formatted string for example 
```
fn to_email_string(&self) -> &str {
    format!("{}@{}", self.user_part, self.domain_part).as_str()
}
```
How about returning Cow<str>?

Reply:
Cow is fine then